### PR TITLE
Added the Neo4j provider

### PR DESCRIPTION
**Pull request by Dinnerb0ne**

### CODE REVIEW

File: providers/neo4j.js [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+		return this.db.run(`MATCH (n:${table}) RETURN n`).then(data => !!data.records.length);
+	}
+
+	/**
+	 * Create a collection within a DB. Options may be specfied, refer to Neo4j docs.
+	 * @param {string} table Name of the Collection to crate
+	 * @returns {Promise<*>} Returns a promise containing the created Collection.
+	 */
+	createTable(table) {
+		return this.db.run(`CREATE (n:${table}) RETURN n`).then(data => data.records.map(node => console.log(node._fields[0].identity.low))[0]);
```
Comment (MrJacz) [inline providers/neo4j.js]: 🤔

File: providers/neo4j.js [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+		this.db = driver.session();
+	}
+
+	/**
+	 * Checks if a table exists.
+	 * @param {string} table The name of the table you want to check.
+	 * @returns {Promise<boolean>}
+	 */
+	hasTable(table) {
+		return this.db.run(`MATCH (n:${table}) RETURN n`).then(data => !!data.records.length);
```
Comment (MrJacz) [inline providers/neo4j.js]: use `Boolean()` instead of `!!` for consistency throughout klasa

File: providers/neo4j.js [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+	}
+
+	/**
+	 * Checks if a document from a Collection exists.
+	 * @param {string} table Name of the Collection
+	 * @param {string} id ID of the document
+	 * @returns {Promise<boolean>}
+	 */
+	has(table, id) {
+		return this.db.run(`MATCH (n:${table} {id: {id} }) RETURN n`, { id }).then(data => !!data.records.length);
```
Comment (MrJacz) [inline providers/neo4j.js]: Same here

File: providers/firestore.js [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+		this.db = firebase.firestore();
+	}
+	/**
+	 *Checks if the table/collection exists
+	 * @param {string} table The name of the table/collection to be checked for existance
+	 * @returns {Promise<boolean>}
+	 */
+	hasTable(table) {
+		return this.db.collection(table).get().then(col => !!col.size);
```
Comment (MrJacz) [inline providers/firestore.js]: same here

File: providers/neo4j.js (hunk 1 above)
Comment (Dinnerb0ne) [reply providers/neo4j.js]: ?

File: providers/neo4j.js [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+	/**
+	 * Update or insert a new value to all entries.
+	 * @param {string} table The name of the table.
+	 * @param {string} path The object to remove or a path to update.
+	 * @param {*} newValue The new value for the key.
+    */
+	async updateValue(table, path, newValue) {
+		const keys = await this.getKeys(table);
+		if (util.isObject(path) && typeof newValue === 'undefined') {
+			Promise.all(keys.map(async node => await this.updatebyID(table, node, path)));
```
Comment (kyranet) [inline providers/neo4j.js]:
```javascript
await Promise.all(keys.map(node => this.updateByID(table, node, path)));
```

File: providers/neo4j.js [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+		const object = this.constructFlatObject(util.mergeObjects(data, this.parseUpdateInput(doc)));
+		return this.db.run(`MATCH (n:${table} {id : {id} }) SET ${object} RETURN n`, { id });
+	}
+
+	/**
+	 * Update or insert a new value to all entries.
+	 * @param {string} table The name of the table.
+	 * @param {string} path The object to remove or a path to update.
+	 * @param {*} newValue The new value for the key.
+    */
```
Comment (kyranet) [inline providers/neo4j.js]: Indentation

File: providers/neo4j.js [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+module.exports = class extends Provider {
+
+	constructor(...args) {
+		super(...args);
+		this.db = null;
+	}
+
+	init() {
+		const driver = neo4j.driver('bolt://localhost', neo4j.auth.basic('neo4j', 'test'));
```
Comment (kyranet) [inline providers/neo4j.js]: This should be configurable via `client.options.providers.neo4j`

File: providers/neo4j.js [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+		return this.db.run(`MATCH (n:${table}) RETURN n`).then(data => Boolean(data.records.length));
+	}
+
+	/**
+	 * Create a collection within a DB. Options may be specfied, refer to Neo4j docs.
+	 * @param {string} table Name of the Collection to crate
+	 * @returns {Promise<*>} Returns a promise containing the created Collection.
+	 */
+	createTable(table) {
+		return this.db.run(`CREATE (n:${table}) RETURN n`).then(data => data.records.map(node => node._fields[0].identity.low))[0];
```
Comment (kyranet) [inline providers/neo4j.js]: You have `[0]` in the wrong place. Also, just do this

```javascript
return this.db.run(`CREATE (n:${table}) RETURN n`).then(data => data.records[0]_fields[0].identity.low);
```

File: providers/neo4j.js [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+	}
+
+	/**
+	 * Retrieves a single Document from a Collection that matches a user determined ID
+	 * @param {string} table Name of the Collection
+	 * @param {string|Object} id ID of the document
+	 * @returns {Promise<*>}
+	 */
+	get(table, id) {
+		return this.db.run(`MATCH (n:${table} {id: {id} }) RETURN n`, { id }).then(data => data.records.map(node => this.packData(node._fields[0].properties, node._fields[0].identity.low))[0]);
```
Comment (kyranet) [inline providers/neo4j.js]: Same here (don't use map, use [0])

File: providers/neo4j.js (hunk 5 above)
Comment (kyranet) [reply providers/neo4j.js]: And same two lines below.

File: providers/neo4j.js [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+	}
+	/**
+	 * Remove a value or object from all entries.
+	 * @param {string} table The name of the table.
+	 * @param {string} path The object to remove or a path to update.
+	 */
+	async removeValue(table, path) {
+		const keys = await this.getKeys(table);
+		if (util.isObject(path) && typeof newValue === 'undefined') {
+			Promise.all(keys.map(node => this.remove(table, node, path)));
```
Comment (kyranet) [inline providers/neo4j.js]: Same here

File: providers/neo4j.js [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+		} else {
+			throw new TypeError(`Expected an object as first parameter or a string and a non-undefined value. Got: ${typeof key} and ${typeof value}`);
+		}
+	}
+	/**
+	 * Deletes a Document from a Collection that matches a user determined ID *
+	 * @param {string} table Name of the Collection
+	 * @param {string} id ID of the document
+	 * @param {string} path Object to be removed
+     * @private
```
Comment (kyranet) [inline providers/neo4j.js]: Indentation

File: providers/neo4j.js [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+	 */
+	async remove(table, id, path) {
+		const [entry] = this.db.run(`MATCH (n:${table} {id: {id} }) RETURN n`, { id }).then(data => data.records.map(node => node._fields[0].properties));
+		delete entry[[Object.keys(path)]];
```
Comment (kyranet) [inline providers/neo4j.js]: This doesn't seem valid to me... and this method's description describes what `delete()` does

File: providers/neo4j.js [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+		return this.db.run(`MATCH (n:${table}) WHERE ID(n) = ${id} SET ${object} RETURN n`);
+	}
+
+	/**
+     * Flattes the Object into a string, to be parsed by Cypher.
+     * @private
+     * @param {Object} object The Object to be flattened
+     * @returns {string}
+     * @private
+     */
```
Comment (kyranet) [inline providers/neo4j.js]: Indentation

File: providers/neo4j.js [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+	constructFlatObject(object) {
+		const { length } = Object.keys(object);
```
Comment (kyranet) [inline providers/neo4j.js]: This method can be refactored away with this:

```javascript
Object.entries(object).map(([key, value]) => `n.${key} = ${typeof value === 'string' ? `'${value}'` : value}`).join(', ');
```

File: providers/neo4j.js [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+		let string = '', count = 0;
+		for (let [key, value] of Object.entries(object)) { // eslint-disable-line
+			if (util.isObject(value)) value = JSON.stringify(value).replace(/"([^"]+)":/g, '$1:');
+			string += `n.${key} = ${typeof string === 'string' ? `'${value}'` : value}${length - count++ === 1 ? '' : ','}`;
+		}
+		return string;
+	}
+
+	/**
+     * Packs the data into easy-to-use manner
```
Comment (kyranet) [inline providers/neo4j.js]: Indentation

File: providers/neo4j.js (hunk 8 above)
Comment (Dinnerb0ne) [reply providers/neo4j.js]: It cant be accessed like an array.

File: providers/neo4j.js (hunk 13 above)
Comment (Dinnerb0ne) [reply providers/neo4j.js]: ?

Comment (Dinnerb0ne) [reply providers/neo4j.js]: I don't see anything out of place.

**Pull request closed without merging**